src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataSerializerHelper.cs

```diff
@@ -0,0 +1,109 @@
+
+            if (instanceAnnotationContainer != null)
+            {
+                IDictionary<string, object> clrAnnotations = instanceAnnotationContainer.GetResourceAnnotations();
+
+                if (clrAnnotations != null)
+                {
+                    foreach (KeyValuePair<string, object> annotation in clrAnnotations)
+                    {
+                        AddODataAnnotations(resource.InstanceAnnotations, resourceContext, annotation, SerializerProvider);
+                    }
+                }
+
+                if (resource.Properties != null)
+                {
+                    foreach (ODataProperty property in resource.Properties)
+                    {
+                        string propertyName = property.Name;
+
+                        if (property.InstanceAnnotations == null)
+                        {
+                            property.InstanceAnnotations = new List<ODataInstanceAnnotation>();
+                        }
+
+                        IDictionary<string, object> propertyAnnotations = instanceAnnotationContainer.GetPropertyAnnotations(propertyName);
+
+                        if (propertyAnnotations != null)
+                        {
+                            foreach (KeyValuePair<string, object> annotation in propertyAnnotations)
+                            {
+                                AddODataAnnotations(property.InstanceAnnotations, resourceContext, annotation, SerializerProvider);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        internal static void AddODataAnnotations(ICollection<ODataInstanceAnnotation> InstanceAnnotations, ResourceContext resourceContext, KeyValuePair<string, object> annotation, ODataSerializerProvider SerializerProvider)
+        {
+            ODataValue annotationValue = null;
+
+            if (annotation.Value != null)
+            {
+                IEdmTypeReference edmTypeReference = resourceContext.SerializerContext.GetEdmType(annotation.Value,
+                    annotation.Value.GetType());
+
+                ODataEdmTypeSerializer edmTypeSerializer = GetEdmTypeSerializer(edmTypeReference, SerializerProvider);
+
+                if (edmTypeSerializer != null)
+                {
+                    annotationValue = edmTypeSerializer.CreateODataValue(annotation.Value, edmTypeReference, resourceContext.SerializerContext);
+                }
+            }
+            else
+            {
+                annotationValue = new ODataNullValue();
+            }
+
+            if (annotationValue != null)
+            {
+                InstanceAnnotations.Add(new ODataInstanceAnnotation(annotation.Key, annotationValue));
+            }
+        }
+
+        private static ODataEdmTypeSerializer GetEdmTypeSerializer(IEdmTypeReference edmTypeReference, ODataSerializerProvider SerializerProvider)
```

Comment:
Is `ODataSerializerProvider.GetEdmTypeSerializer()` not sufficient?

Reply:
It is. This is just a helper method that calls the appropriate `EdmTypeSerializer` depending on the the 'EdmTypeReference' being serialized - if it is a `collection`, `structured type`, or none of the above.